activation/handler_v2.go

```diff
@@ -0,0 +1,177 @@
+		return nil, fmt.Errorf("failed to check if atx exists: %w", err)
+	}
+	if exists {
+		return nil, nil
+	}
+
+	h.log.Debug(
+		"processing atx",
+		log.ZContext(ctx),
+		zap.Stringer("atx_id", watx.ID()),
+		zap.Uint32("publish", watx.PublishEpoch.Uint32()),
+		zap.Stringer("smesherID", watx.Smesher()),
+	)
+
+	if err := h.syntacticallyValidate(ctx, watx); err != nil {
+		return nil, fmt.Errorf("atx %s syntactically invalid: %w", watx.ID(), err)
+	}
+
+	// TODO:
+	// 1. fetch dependencies
+	// 2. syntactically validate dependencies
+	// 3. contextually validate ATX
+	// 4. store ATX
+	return nil, nil
+}
+
+// Syntactically validate an ATX.
+// TODOs:
+// 1. support marriages
+// 2. support merged ATXs.
+func (h *HandlerV2) syntacticallyValidate(ctx context.Context, atx *wire.ActivationTxV2) error {
+	if !h.edVerifier.Verify(signing.ATX, atx.Smesher(), atx.SignedBytes(), atx.Signature) {
+		return fmt.Errorf("invalid atx signature: %w", errMalformedData)
+	}
+	if atx.PositioningATX == types.EmptyATXID {
+		return errors.New("empty positioning atx")
+	}
+	// TODO: support marriages
+	if len(atx.Marriages) != 0 {
+		return errors.New("marriages are not supported")
+	}
+
+	current := h.clock.CurrentLayer().GetEpoch()
+	if atx.PublishEpoch > current+1 {
+		return fmt.Errorf("atx publish epoch is too far in the future: %d > %d", atx.PublishEpoch, current+1)
+	}
+
+	if atx.MarriageATX == nil {
+		if len(atx.NiPosts) != 1 {
+			return errors.New("solo atx must have one nipost")
+		}
+		if len(atx.NiPosts[0].Posts) != 1 {
+			return errors.New("solo atx must have one post")
+		}
+		if atx.NiPosts[0].Posts[0].PrevATXIndex != 0 {
+			return errors.New("solo atx post must have prevATXIndex 0")
+		}
+	}
+
+	if atx.Initial != nil {
+		if atx.MarriageATX != nil {
+			return errors.New("initial atx cannot reference a marriage atx")
+		}
+		if atx.Initial.CommitmentATX == types.EmptyATXID {
+			return errors.New("initial atx missing commitment atx")
+		}
+		if atx.VRFNonce == nil {
+			return errors.New("initial atx missing vrf nonce")
+		}
+		if len(atx.PreviousATXs) != 0 {
+			return errors.New("initial atx must not have previous atxs")
+		}
+
+		if atx.Coinbase == nil {
+			return errors.New("initial atx missing coinbase")
+		}
+
+		numUnits := atx.NiPosts[0].Posts[0].NumUnits
+		if err := h.nipostValidator.VRFNonceV2(
+			atx.Smesher(), atx.Initial.CommitmentATX, types.VRFPostIndex(*atx.VRFNonce), numUnits,
+		); err != nil {
+			return fmt.Errorf("invalid vrf nonce: %w", err)
+		}
+		if err := h.nipostValidator.PostV2(
+			ctx, atx.Smesher(), atx.Initial.CommitmentATX, &atx.Initial.Post, shared.ZeroChallenge, numUnits,
+		); err != nil {
+			return fmt.Errorf("invalid initial post: %w", err)
+		}
+		return nil
+	}
+
+	for i, prev := range atx.PreviousATXs {
+		if prev == types.EmptyATXID {
+			return fmt.Errorf("previous atx[%d] is empty", i)
+		}
+		if prev == h.goldenATXID {
+			return fmt.Errorf("previous atx[%d] is the golden ATX", i)
+		}
```

Comment:
I think we shouldn't explicitly check for those two, but rather if the referenced Previous ATX belongs to the SmesherID we expect?

Also I think we should do this after the next check - to stop validating earlier if we have an ATX with no `MarriageATX` but multiple PreviousATXs.

Reply:
> I think we shouldn't explicitly check for those two, but rather if the referenced Previous ATX belongs to the SmesherID we expect?

Checking the owner of the previous ATXs requires fetching them. It's part of #5949. In "syntactical validation **w/o** deps" we want to check as much as possible to reject invalid ATXs before downloading the dependencies.

>Also I think we should do this after the next check - to stop validating earlier if we have an ATX with no MarriageATX but multiple PreviousATXs.

It could be moved :+1: but I don't expect an observable perf difference :)

Reply:
Well, we don't want to fetch 256 previous ATXs before checking if a `MarriageATX` is present in the ATX 😉